### PR TITLE
[runtime env] support create or delete runtime envs in agent

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

- Support create or delete runtime env from raylet to agent.
- The new rpc is triggered by PopWorker of worker_pool.
- Add uris to serialized_runtime_env of task spec to ensure runtime env creating could be triggered.
- This PR is not a complete implementation. Some future issues:
    * Try `ensure_runtime_env_setup(uris)` to download packages in agent.
    * Delete runtime env local files in agent.
    *  Reschedule to other nodes when create runtime env failed. We can make PopWorker to be a async function.

## Related issue number
https://github.com/ray-project/ray/issues/15696

### CODE REVIEW

File: src/ray/raylet/worker_pool.cc
Comment (edoakes) [inline src/ray/raylet/worker_pool.cc:899]: This will be a pretty bad user experience if there is an error in the environment that causes it to continually fail. Ideally, we should optionally retry some number of times and then fail the tasks that depend on the environment with an `EnvironmentFailureError` or something

**Comment by edoakes** [comment]
@SongGuyang please merge latest master, @architkulkarni recently update the worker caching logic in worker_pool.cc to properly handle runtime_envs

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc]: @edoakes Yes, I have add some retry code.

**Comment by SongGuyang** [comment]
@edoakes Code has been updated. Please help to review. @architkulkarni @iycheng

**Comment by edoakes** [comment]
@SongGuyang we are pushing for an internal deadline this coming Monday so we might be slower than usual to respond to PRs. I will let @architkulkarni @iycheng review this when they have a chance!

**Comment by SongGuyang** [comment]
@edoakes OK, no problem.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:951]: Why don't we need to add to and erase from `state.tasks_to_pending_runtime_envs` like we did in this part? https://github.com/ray-project/ray/pull/15904/files#diff-4c19d6ebd8575189e9439fa5646b567546dc54f64d576a959ce1bb43902d1a03R882-R887

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:961]: Thanks for adding this!

File: python/ray/_private/runtime_env.py
Comment (iycheng) [inline python/ray/_private/runtime_env.py:117]: make it python style: `set_uris`

File: src/ray/raylet/agent_manager.h
Comment (iycheng) [inline src/ray/raylet/agent_manager.h:96]: Why shared pointer here? Btw, before this change, it's a reference unique ptr which is also not good. Either a reference or a raw ptr should be good here.

File: src/ray/raylet/node_manager.cc
Comment (iycheng) [inline src/ray/raylet/node_manager.cc:324]: agent_manager_ = make_unqiue<AgentManager>(...)

**Comment by iycheng** [comment]
I'm thinking about the cleanup step introduced by the agent manager. Right now the cleanup is carefully designed to be:
- if there is no referencing to some object, move it and delete it in another process (util worker before). The reason we did mv first is to avoid race conditions.
- and the downloading happens after the connection, which is also to avoid race condition. it increases the counting by 1 first (like locking) and then downloading. 

I'm wondering will we follow the same pattern here? or we'll introduce something better? If downloading and uploading happens both in agent, should the agent manage the referencing of the uris?

File: src/ray/raylet/agent_manager.h
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.h:96]: Maybe I can redefine to `std::shared_ptr<AgentManager> delegate_`? I use shared_ptr because I should access AgentManager in worker_pool.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:28]: ok

File: src/ray/common/ray_config_def.h
Comment (SongGuyang) [reply src/ray/common/ray_config_def.h:318]: Unrelated. This timeout is a retry interval when agent connection is unavailable. By the way, the agent will send reply when runtime env installation finished.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:951]: Because this worker is not a dedicated worker which bound to a task. If we put it to `tasks_to_pending_runtime_envs`, it could not be reused.

File: src/ray/raylet/agent_manager.h
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.h:96]: I know the reason why a ptr reference before. It is because  the construction of `DefaultAgentManagerServiceHandler` is happened before `AgentManger` construction.

**Comment by SongGuyang** [comment]
@iycheng Maybe agent should manage referencing. But raylet should trigger the increase and decrease of references when job stopped or actor killed.

**Review by architkulkarni — APPROVED** [review]
Looks good to me!  I don't have merge privileges so I'll let @edoakes handle it from here

**Comment by SongGuyang** [comment]
@architkulkarni Thanks. @edoakes Can we merge this?

**Review by edoakes — COMMENTED** [review]
@SongGuyang it looks like we don't have any tests for this, could you add some basic tests before we merge? If possible, it'd be ideal if we had a feature flag and ran the existing test with the feature flag on and off but it seems that will require implementing more features. In that case just some basic unit tests are fine.

**Comment by SongGuyang** [comment]
@edoakes Sorry for late update. I have updated some unit tests to `worker_pool_test`. And I think we can trust `test_runtime_env.py` to be the integration testing because this testing will be failed when the runtime env agent rpc failed now. But we need some other PRs to make this feature completed. So we will add more test code in future PRs.
Please help to review and merge it if it look good for you. Thank you.

**Comment by architkulkarni** [comment]
Thanks so much for adding these tests!  We are currently pushing for the Ray 1.4 release so our review may be slower than usual again--sorry about that.  We'll try to review it as soon as we can

**Comment by SongGuyang** [comment]
@architkulkarni OK. Thanks.

File: src/ray/raylet/agent_manager.cc
Comment (raulchen) [inline src/ray/raylet/agent_manager.cc:106]: 1) `serialized_runtime_env` is binary data? Is this human readable?
2) This may be clearer: `Runtime env agent is not registered yet. Will retry CreateRuntimeEnv later.`

File: src/ray/raylet/agent_manager.cc
Comment (raulchen) [inline src/ray/raylet/agent_manager.cc:109]: Instead of retrying with a delay, we can also save the request in a list.

File: src/ray/raylet/agent_manager.cc
Comment (raulchen) [inline src/ray/raylet/agent_manager.cc:130]: why reset the connection? Will the agent re-register to the raylet?

File: src/ray/raylet/node_manager.cc
Comment (raulchen) [inline src/ray/raylet/node_manager.cc:335]: 1) why will the ip or the port be empty? Is this only for tests?
2) Make the log clearer.

File: src/ray/raylet/worker_pool.cc
Comment (raulchen) [inline src/ray/raylet/worker_pool.cc:880]: Add a `HasRuntimeEnv` method in task spec?

File: src/ray/raylet/agent_manager.h
Comment (raulchen) [inline src/ray/raylet/agent_manager.h:99]: why reference of shared_ptr, not just shared_ptr?

File: src/ray/raylet/agent_manager.cc
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.cc:106]: `serialized_runtime_env` is a readable json string now.

File: src/ray/raylet/agent_manager.cc
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.cc:130]: Sorry, I have checked agent code just now. No re-register now in agent. We should delete this "reset" and reserve this grpc client.

File: src/ray/raylet/node_manager.cc
Comment (SongGuyang) [reply src/ray/raylet/node_manager.cc:335]: Maybe no ip or port in agent register message.

Comment (SongGuyang) [reply src/ray/raylet/node_manager.cc:335]: Or in tests.

Comment (SongGuyang) [reply src/ray/raylet/node_manager.cc:335]: Maybe I should use RAY_CHECK instead.

File: src/ray/raylet/agent_manager.h
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.h:99]: I just use shared_ptr instead of unique_ptr. 
 It is because the construction of DefaultAgentManagerServiceHandler is happened before AgentManger construction.

File: dashboard/agent.py
Comment (raulchen) [inline dashboard/agent.py:295]: `runtime_env_dir`?

File: dashboard/agent.py
Comment (raulchen) [inline dashboard/agent.py:299]: used

File: src/ray/common/ray_config_def.h
Comment (raulchen) [inline src/ray/common/ray_config_def.h:330]:
```
/// If the agent manager fails to communicate with the dashboard agent, we will retry after this interval.
RAY_CONFIG(uint32_t, agent_manager_retry_interval_ms, 1000);
```

File: src/ray/protobuf/runtime_env_agent.proto
Comment (raulchen) [inline src/ray/protobuf/runtime_env_agent.proto:27]: `job env` -> `runtime env`

File: src/ray/raylet/agent_manager.cc
Comment (raulchen) [inline src/ray/raylet/agent_manager.cc:171]: `DeleteRuntimeEnv`

File: src/ray/raylet/agent_manager.cc
Comment (raulchen) [inline src/ray/raylet/agent_manager.cc:168]: `will retry it later`

File: src/ray/raylet/worker_pool.cc
Comment (raulchen) [inline src/ray/raylet/worker_pool.cc:959]: `WarnAboutSize` is actually used to print a warning when starting too many workers. we should move it into the `start_worker_process_fn` callback, after `StartWorkerProcess`.

File: src/ray/raylet/agent_manager.h
Comment (raulchen) [inline src/ray/raylet/agent_manager.h:38]: `done` -> `successful`, and document this.

File: src/ray/raylet/worker_pool.cc
Comment (raulchen) [inline src/ray/raylet/worker_pool.cc:948]: This log doesn't seem necessary. We've logged about the failure in the agent manager.

File: src/ray/raylet/worker_pool.h
Comment (raulchen) [inline src/ray/raylet/worker_pool.h:466]: The value is not used, we can just use a set instead of a map: ` std::unordered_set<TaskID> tasks_with_pending_runtime_envs;`
By the way, could you also change `tasks_to_dedicated_workers` to a set?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (raulchen) [inline dashboard/modules/runtime_env/runtime_env_agent.py:30]: `Creating`

File: src/ray/raylet/agent_manager.h
Comment (raulchen) [inline src/ray/raylet/agent_manager.h:39]: Does this callback need the done flag?

**Comment by architkulkarni** [comment]
I reviewed the test, and it looks great, thanks for adding it @SongGuyang!  It looks like you've also tested that the runtime env hashing works as expected, and your framework will make it easy for me to add more tests later.  I agree that the existing Python tests can serve as integration testing.  Thanks also @raulchen for the careful review.

This PR looks good to me, but perhaps we can also get an approval from @iycheng or @edoakes who might have more experience with gRPC?

File: src/ray/raylet/worker_pool.h
Comment (raulchen) [inline src/ray/raylet/worker_pool.h:463]: `tasks_with_dedicated_workers`. The comment needs to be changed as well. It can be something like "All tasks that have associated dedicated workers.".

File: src/ray/raylet/worker_pool.h
Comment (raulchen) [inline src/ray/raylet/worker_pool.h:465]: All tasks that have pending runtime envs.

**Review by raulchen — APPROVED** [review]

**Comment by raulchen** [comment]
HI, @architkulkarni the gRPC changes are totally fine. They are just boilerplate code. 
I've carefully reviewed this PR a few times. And it should be good to merge. I'm going to merge it after CI passes to unblock the follow-up PRs (@SongGuyang already has a new PR that depends on this one).

**Pull request merged**
